homeassistant/components/private_ble_device/coordinator.py

```diff
@@ -0,0 +1,270 @@
+"""Central manager for tracking devices with random but resolvable MAC addresses."""
+from __future__ import annotations
+
+import binascii
+from collections.abc import Callable
+import logging
+from typing import cast
+
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+
+from homeassistant.components import bluetooth
+from homeassistant.components.bluetooth.match import BluetoothCallbackMatcher
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PADDING = b"\x00" * 13
+
+UnavailableCallback = Callable[[bluetooth.BluetoothServiceInfoBleak], None]
+Cancellable = Callable[[], None]
+
+
+def _async_get_cipher_for_irk(irk: bytes) -> Cipher:
+    return Cipher(algorithms.AES(irk), modes.ECB())
+
+
+def _async_service_info_matches_irk(
+    service_info: bluetooth.BluetoothServiceInfoBleak, cipher: Cipher
+) -> bool:
+    # https://www.mdpi.com/2227-7390/10/22/4346
+    # Use 128bit IRK as encryption key for ECB AES.
+    # Encrypt prand (the MSB, top 24bits of MAC address) with the IRK
+    # It's zero padded to 128bits.
+    # The top 24bits of the cipher text are taken as a hash
+    # This hash should match the low 24bits of the MAC address.
+    rpa = binascii.unhexlify(service_info.address.replace(":", ""))
+
+    if rpa[0] & 0xC0 != 0x40:
+        # Not an RPA
+        return False
+
+    pt = PADDING + rpa[:3]
+
+    encryptor = cipher.encryptor()
+    ct = encryptor.update(pt) + encryptor.finalize()
+
+    if ct[13:] != rpa[3:]:
+        return False
+
+    return True
```

Comment:
maybe move this code into a PyPI pkg

Reply:
Was going to talk to you about how to split it up. "Is this an RPA?" Felt like it belonged in the Bluetooth component, perhaps even a filter for it when grabbing all service infos?

Wasn't sure there was enough else to justify a separate package. It would just be creating a cipher from a string.

Reply:
Could be added to https://github.com/Bluetooth-Devices/bluetooth-data-tools